spam/03_spam_data_slicing_tutorial.py

```diff
@@ -0,0 +1,483 @@
+# We'll assume that you are familiar with Snorkel's the data/model/training abstraction — if not, we'd recommend you check out our [MTL Tutorial](https://github.com/snorkel-team/snorkel-tutorials/blob/master/mtl/multitask_tutorial.ipynb).
+
+# %% [markdown]
+# ### Featurize Data
+#
+# As a first step, we'll featurize the data—as you saw in the introductory Spam tutorial, we'll extract simple bag of words features and store them as numpy arrays.
+
+# %%
+import torch
+from sklearn.feature_extraction.text import CountVectorizer
+from snorkel.classification.data import DictDataset, DictDataLoader
+
+vectorizer = CountVectorizer(ngram_range=(1, 1))
+
+
+def df_to_torch_example(vectorizer, df, fit_train=False):
+    words = [row.text for i, row in df.iterrows()]
+
+    if fit_train:
+        feats = vectorizer.fit_transform(words)
+    else:
+        feats = vectorizer.transform(words)
+    X = feats.todense()
+    Y = df["label"].values
+    return X, Y
+
+
+# %%
+X_train, Y_train = df_to_torch_example(vectorizer, df_train, fit_train=True)
+X_valid, Y_valid = df_to_torch_example(vectorizer, df_valid, fit_train=False)
+X_test, Y_test = df_to_torch_example(vectorizer, df_test, fit_train=False)
+
+# %% [markdown]
+# ### Create DataLoaders
+#
+# Next, we'll use the extracted Tensors to initialize a `DictDataLoader` — as a quick recap, this is a Snorkel-specific class that inherits from the common PyTorch class and supports multiple data fields in the `X_dict` and labels in the `Y_dict`.
+#
+# In this task, we'd like to store the `bow_features` in our `X_dict`, and we have one set of labels (for now) correpsonding to the `spam_task`.
+
+# %%
+BATCH_SIZE = 32
+
+
+def create_dict_dataloader(X, Y, split, **kwargs):
+    ds = DictDataset(
+        name="spam_dataset",
+        split=split,
+        X_dict={"bow_features": torch.FloatTensor(X)},
+        Y_dict={"spam_task": torch.LongTensor(Y)},
+    )
+    return DictDataLoader(ds, **kwargs)
+
+
+dl_train = create_dict_dataloader(
+    X_train, Y_train, split="train", batch_size=BATCH_SIZE, shuffle=True
+)
+dl_valid = create_dict_dataloader(
+    X_valid, Y_valid, split="valid", batch_size=BATCH_SIZE, shuffle=False
+)
+dl_test = create_dict_dataloader(
+    X_test, Y_test, split="test", batch_size=BATCH_SIZE, shuffle=False
+)
+
+# %% [markdown]
+# We can inspect our datasets to confirm that they have the appropriate fields.
+
+# %%
+dl_valid.dataset
+
+# %% [markdown]
+# ### Define `SnorkelClassifier`
+#
+# We'll define a simple Multi-Layer Perceptron (MLP) architecture to learn from the `bow_features`.
+#
+# _Note: the following might feel like extra steps to define what is a very simple architecture, but this will lend us additional flexibility later in the pipeline!_
+
+# %% [markdown]
+# To start, we define a `module_pool` with all the [PyTorch](https://pytorch.org) modules that we'll want to include in our network.
```

Comment:
we'll -> we. In general, I think we can speak more present tense throughout the tutorial (mostly just less of "we'll")